crates/ol/stf/src/tests_snark.rs

```diff
@@ -217,3 +402,567 @@ fn test_snark_account_deposit_and_withdrawal() {
+        1,
+        "Inbox should have 2 messages (deposit + GAM)"
+    );
+
+    // Balance unchanged (GAM messages have 0 value)
+    assert_eq!(
+        snark_account.balance(),
+        BitcoinAmount::from_sat(100_000_000),
+        "Snark account balance should be unchanged"
+    );
+}
+
+fn get_snark_state_expect(
+    state: &OLState,
+    snark_id: AccountId,
+) -> (&NativeAccountState, &NativeSnarkAccountState) {
+    let snark_account = state.get_account_state(snark_id).unwrap().unwrap();
+    (snark_account, snark_account.as_snark_account().unwrap())
+}
+
+#[test]
+fn test_snark_update_process_inbox_message_with_valid_proof() {
+    let mut state = OLState::new_genesis();
+    let snark_id = test_account_id(100);
+    let recipient_id = test_account_id(200);
+
+    // Setup: genesis with snark account
+    let (genesis_header, _snark_serial) =
+        setup_genesis_with_snark_account(&mut state, snark_id, 100_000_000);
+
+    // Create recipient account
+    create_empty_account(&mut state, recipient_id);
+
+    // Create parallel MMR tracker to generate proofs
+    let mut inbox_tracker = InboxMmrTracker::new();
+
+    // Step 1: Send a message to snark account inbox
+    let msg_data = vec![1u8, 2, 3, 4];
+    let gam_tx = TransactionPayload::GenericAccountMessage(
+        GamTxPayload::new(snark_id, msg_data.clone()).expect("Should create GAM payload"),
+    );
+    let (slot, epoch) = (1, 0);
+    execute_tx_in_block(&mut state, &genesis_header, gam_tx, slot, epoch)
+        .expect("GAM should succeed");
+
+    // Track the message in parallel MMR (must match exactly what was inserted)
+    let gam_msg_entry = MessageEntry::new(
+        crate::constants::SEQUENCER_ACCT_ID,
+        epoch, // epoch when message was added
+        MsgPayload::new(BitcoinAmount::from_sat(0), msg_data),
+    );
+
+    // Debug: compute hashes both ways to see the mismatch
+    {
+        use strata_merkle::hasher::MerkleHasher;
+        use tree_hash::TreeHash;
+
+        let tree_hash_result =
+            <MessageEntry as TreeHash>::tree_hash_root(&gam_msg_entry).into_inner();
+        let msg_bytes: Vec<u8> = gam_msg_entry.as_ssz_bytes();
+        let hash_leaf_result = StrataHasher::hash_leaf(&msg_bytes);
+
+        eprintln!("DEBUG: TreeHash result    = {:?}", tree_hash_result);
+        eprintln!("DEBUG: hash_leaf result   = {:?}", hash_leaf_result);
+        eprintln!(
+            "DEBUG: Hashes match? {}",
+            tree_hash_result == hash_leaf_result
+        );
+    }
+
+    let gam_proof = inbox_tracker.add_message(&gam_msg_entry);
+
+    // Step 2: Verify the parallel MMR matches the actual inbox MMR
+    let snark_account = state.get_account_state(snark_id).unwrap().unwrap();
+    let snark_state = snark_account.as_snark_account().unwrap();
+
+    assert_eq!(
+        snark_state.inbox_mmr().num_entries(),
+        inbox_tracker.num_entries(),
+        "Parallel MMR must stay synchronized with actual inbox MMR"
+    );
```

Comment:
we only verify that the number of entries is equal, right? shouldn't match items/proofs?